### PR TITLE
[MM-55726] Create batch report worker, add batch report job for exporting users to CSV

### PR DESCRIPTION
**Pull request by devinbinnie**
#### Summary
This PR contains changes to allow an admin to export users to CSV.

A quick breakdown:
- Refactored the `BatchMigrationWorker` used for migrations into the generic `BatchWorker`, and extended a module called `BatchReportWorker`, which will get and save report data in chunks, until no more rows are available, then will compile the chunks and send a DM to the requesting user to pick up the report from the file store.
- Created the `export_users_to_csv` job that provides the method for getting data, and updates parameters for the job accordingly.
- Exposed two APIs: one for starting the batch export job specifically for users, with a `format` option (currently only `csv` works), and another for picking up the resulting report by passing the job ID (ie. the report ID) as a parameter.

Still to come in future changes:
- We could use a better way to display the DM. Currently we just have it set to write a message with the link (which doesn't work unless you tell Markdown to ignore the path). A separate component for this would be a good idea.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55726

-->
```release-note
NONE
```

### CODE REVIEW

**Comment by devinbinnie** [comment]
Looks like CodeQL is complaining about my use of the file store where the user provides part of the file name that is used (ie. the format extension)

This is validated in the API using a function called `IsValidReportExportFormat`, but seems that CodeQL doesn't pick it up. Is there a way I can hint at it to clear that error? Or should I be using some other constant to provide the extension?

**Review by harshilsharma63 — CHANGES_REQUESTED** [review]

File: server/channels/app/report.go [hunk 1]
```diff
@@ -4,11 +4,131 @@
...
+		channel, err := a.GetOrCreateDirectChannel(request.EmptyContext(a.Log()), userID, systemBot.UserId)
+		if err != nil {
+			rctx.Logger().Error("Failed to post batch export finished message", mlog.Err(err))
+			return
+		}
+
+		post := &model.Post{
+			ChannelId: channel.Id,
+			Message:   i18n.T("app.report.send_report_to_user.export_finished", map[string]string{"Link": a.GetSiteURL() + "/api/v4/reports/export/" + jobId + "?format=" + format}),
+			Type:      model.PostTypeAdminReport,
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/report.go:104]: This will likely be changing on we have UX look at it. We'll likely want a custom component of some kind.

File: server/channels/app/report.go [hunk 2]
```diff
@@ -30,3 +150,74 @@ func (a *App) getUserReport(filter *model.UserReportOptions) ([]*model.UserRepor
...
+	}
+
+	// Check for existing job
+	// TODO: Maybe make this a reusable function?
+	pendingJobs, err := a.Srv().Jobs.GetJobsByTypeAndStatus(rctx, model.JobTypeExportUsersToCSV, model.JobStatusPending)
+	if err != nil {
+		return err
+	}
+	for _, job := range pendingJobs {
+		if job.Data["date_range"] == dateRange && job.Data["requesting_user_id"] == rctx.Session().UserId {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/report.go:167]: I suppose we could add the other user ids to the job data, but there are some nuances here we'd have to account for:
- What if they request the report right as the job is sending before it's been marked Done? Then the user wouldn't get anything.
- Depending on the length of time for the report to generate, what if a new user was created that the second user who asked for the report expected to see? The already generated chunks wouldn't contain that user.

I think we're over-optimizing here for a case that likely wouldn't have a huge affect on the system and while I do like the idea, I think we should revisit it elsewhere since we'd need to think it through a bit more.

File: server/channels/app/report.go [hunk 3]
```diff
@@ -30,3 +150,74 @@ func (a *App) getUserReport(filter *model.UserReportOptions) ([]*model.UserRepor
...
+			return model.NewAppError("StartUsersBatchExport", "app.report.start_users_batch_export.job_exists", nil, "", http.StatusBadRequest)
+		}
+	}
+
+	inProgressJobs, err := a.Srv().Jobs.GetJobsByTypeAndStatus(rctx, model.JobTypeExportUsersToCSV, model.JobStatusInProgress)
+	if err != nil {
+		return err
+	}
+	for _, job := range inProgressJobs {
+		if job.Data["date_range"] == dateRange && job.Data["requesting_user_id"] == rctx.Session().UserId {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/report.go:177]: See above.

**Comment by devinbinnie** [comment]
@harshilsharma63 Here's the sample file

[batch_report_pmwekr4jbfrhxn9qidpuq6ejrh.csv](https://github.com/mattermost/mattermost/files/13862099/batch_report_pmwekr4jbfrhxn9qidpuq6ejrh.csv)

File: server/channels/app/report.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/channels/app/report.go]: OKay.

File: server/channels/app/report.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/channels/app/report.go:172]: We know for sure this is super duper duper duper heavy on MySQL, so that's why I'm trying to avoid duplicate work. Can we at least try to come up with solution to avoid duplicate data processing?

**Comment by harshilsharma63** [comment]
> @harshilsharma63 Here's the sample file
> 
> [batch_report_pmwekr4jbfrhxn9qidpuq6ejrh.csv](https://github.com/mattermost/mattermost/files/13862099/batch_report_pmwekr4jbfrhxn9qidpuq6ejrh.csv)

I don't know if this is due to dummy data or what, but the data in these doesn't seem correct. I'm pasting a few rows here for quick reference-

|Id                        |Username     |Email                    |CreateAt     |Name       |Roles|LastLogin                        |LastStatusAt|LastPostDate|DaysActive|TotalPosts|
|--------------------------|-------------|-------------------------|-------------|-----------|-----|---------------------------------|------------|------------|----------|----------|
|7ks164eoxbrhicymmwenckn6hh|testuser-1   |testuser-1@example.com   |testuser-1   |system_user|     |2023-06-19 22:31:24.445 -0400 EDT|            |2           |33242     |          |
|7591d1y5gff8fbyu7zimbzznew|testuser-10  |testuser-10@example.com  |testuser-10  |system_user|     |2023-06-19 22:31:24.241 -0400 EDT|            |2           |33179     |          |
|kr5ue3ci6bbt38z9new1hi8fma|testuser-100 |testuser-100@example.com |testuser-100 |system_user|     |2023-06-19 22:31:24.021 -0400 EDT|            |2           |32892     |          |
|7rykehwr13b99khnsaa89euncc|testuser-1000|testuser-1000@example.com|testuser-1000|system_user|     |2023-06-22 01:18:43.424 -0400 EDT|            |2           |33596     |          |
|xn56gwetbfr75mxm7475s9d11y|testuser-1001|testuser-1001@example.com|testuser-1001|system_user|     |2023-06-22 01:18:43.705 -0400 EDT|            |2           |33368     |          |
|a313fikyb3bzffb4a5oa5cyfua|testuser-1002|testuser-1002@example.com|testuser-1002|system_user|     |2023-06-22 01:18:42.89 -0400 EDT |            |2           |33390     |          |

The values in `CreateAt`, `Name`, `LastPostDate` and `TotalPost` seem incorrect.

Seems like the values in create at are actually names, and the values in names are actually roles. 

Last post date is a number, which doesn't signify anything IMO, shouldn't it be a calendar date like `5 Jan 2022` or something? Because the report can span multiple months and years.

The empty total posts seems incorrect. If there is a last post date, there should definitely be a total post count.

Also, I think we should add a couple of rows at the start to include report metadata like who generated it, what params were used, what time period etc, you know, like a bank statement. Where if I send the report, the receiver doesn't need to ask me all that, its just right in there, in the report. Thoughts @johndavidlugtu @abhijit-singh ?

**Comment by devinbinnie** [comment]
> Seems like the values in create at are actually names, and the values in names are actually roles.
> Last post date is a number, which doesn't signify anything IMO, shouldn't it be a calendar date like `5 Jan 2022` or something? Because the report can span multiple months and years.
> The empty total posts seems incorrect. If there is a last post date, there should definitely be a total post count.

I think there's a column that's not being exported correctly. Will have a look.

> Also, I think we should add a couple of rows at the start to include report metadata like who generated it, what params were used, what time period etc, you know, like a bank statement. Where if I send the report, the receiver doesn't need to ask me all that, its just right in there, in the report. Thoughts @johndavidlugtu @abhijit-singh ?

This isn't generally how a CSV export would work. CSV is usually just raw data, you wouldn't be adding metadata anywhere since users would want to be able to import it into a table.

If we did some kind of PDF report I could see the value in doing that.

File: server/channels/app/report.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/report.go:172]: I think this has too many moving parts to be done right now, and it's out-of-scope for this PR.
Let's revisit this elsewhere.

File: server/public/model/report.go [hunk 4]
```diff
@@ -75,6 +82,43 @@ type UserReport struct {
...
 }
 
+func (u *UserReport) ToReport() []string {
+	lastStatusAt := ""
+	if u.LastStatusAt != nil {
+		lastStatusAt = time.UnixMilli(*u.LastStatusAt).String()
+	}
+	lastPostDate := ""
+	if u.LastPostDate != nil {
+		lastStatusAt = time.UnixMilli(*u.LastPostDate).String()
```
Comment (harshilsharma63) [inline server/public/model/report.go:92]: Bug: should be assigning to `lastPostDate`, not `lastStatusAt`.

**Review by harshilsharma63 — APPROVED** [review]
Approving but one line change needed.

**Comment by harshilsharma63** [comment]
> This one seems to look better: [batch_report_kyops1quetdauyxhe9w3wztd4w.csv](https://github.com/mattermost/mattermost/files/13875928/batch_report_kyops1quetdauyxhe9w3wztd4w.csv)
> 
> I might have just given you an out-of-date one, sorry about that.

Last post date is still empty. I've added a comment on the line causing this. Otherwise LGTM!

File: server/public/model/report.go (hunk 4 above)
Comment (devinbinnie) [reply server/public/model/report.go:92]: Whoops that'll do it lol

File: server/channels/app/report.go [hunk 5]
```diff
@@ -4,11 +4,136 @@
...
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/platform/shared/filestore"
 )
 
+func (a *App) SaveReportChunk(format string, prefix string, count int, reportData []model.ReportableObject) *model.AppError {
+	switch format {
+	case "csv":
+		return a.saveCSVChunk(prefix, count, reportData)
+	}
```
Comment (lieut-data) [inline server/channels/app/report.go:23]: It seems weird to have support for `format` everywhere (with only one type), but then hard-code the `ReportableObject` with a `ToReport()` that returns a `[]string` that's probably (?) only useful for CSV. It feels like we're pre-designing for a future alternative format but we'll have to change a ton to support that anyway. Would we consider simplifying and removing the `format` handling altogether?

File: server/channels/app/report.go [hunk 6]
```diff
@@ -4,11 +4,136 @@
...
+	w := csv.NewWriter(&buf)
+
+	for _, report := range reportData {
+		err := w.Write(report.ToReport())
+		if err != nil {
+			return model.NewAppError("saveCSVChunk", "app.save_csv_chunk.write_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+	}
+
+	w.Flush()
```
Comment (lieut-data) [inline server/channels/app/report.go:38]: As per [documentation](https://pkg.go.dev/encoding/csv#Writer.Flush), it looks like might need to check for an error from `w.Flush()`:

> To check if an error occurred during the Flush, call Error.

File: server/channels/jobs/batch_report_worker.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+)
+
+type BatchReportWorkerAppIFace interface {
+	SaveReportChunk(format string, prefix string, count int, reportData []model.ReportableObject) *model.AppError
+	CompileReportChunks(format string, prefix string, numberOfChunks int, headers []string) *model.AppError
+	SendReportToUser(rctx request.CTX, userID string, jobId string, format string) *model.AppError
+	CleanupReportChunks(format string, prefix string, numberOfChunks int) *model.AppError
+}
+
+type BatchReportWorker[T BatchReportWorkerAppIFace] struct {
```
Comment (lieut-data) [inline server/channels/jobs/batch_report_worker.go:25]: Why is a generic being used here? Can't `app` just be of type `BatchReportWorkerAppIFace`?

File: server/channels/jobs/batch_report_worker.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	headers []string,
+	getData func(jobData model.StringMap, app T) ([]model.ReportableObject, model.StringMap, bool, error),
+) model.Worker {
+	worker := &BatchReportWorker[T]{
+		app:          app,
+		reportFormat: reportFormat,
+		headers:      headers,
+		getData:      getData,
+	}
+	worker.BatchWorker = BatchWorker{
```
Comment (lieut-data) [inline server/channels/jobs/batch_report_worker.go:48]: nit: Should we inline setting `BatchWorker` above vs. assigning it discretely?

File: server/channels/jobs/batch_report_worker.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+			worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		} else {
+			worker.logger.Info("Worker: Report job complete")
+			worker.setJobSuccess(worker.logger, job)
+		}
+
+		return true
+	}
+
+	err = worker.saveData(job, reportData)
```
Comment (lieut-data) [inline server/channels/jobs/batch_report_worker.go:79]: nit: `getData` and `saveData` feel like analogues, except one is about the job data, and the other actually does the processing. Maybe we call the latter `processChucnk` or some such to help clarify?

File: server/channels/jobs/batch_report_worker.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+		worker.logger.Error("Worker: Failed to save report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	}
+
+	for k, v := range nextData {
+		job.Data[k] = v
+	}
+
+	// TODO add progress?
```
Comment (lieut-data) [inline server/channels/jobs/batch_report_worker.go:90]: Is there a next action for this TODO?

File: server/channels/jobs/batch_report_worker.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+	err = worker.saveData(job, reportData)
+	if err != nil {
+		worker.logger.Error("Worker: Failed to save report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	}
+
+	for k, v := range nextData {
+		job.Data[k] = v
+	}
```
Comment (lieut-data) [inline server/channels/jobs/batch_report_worker.go:88]: It feels safer to `job.Data = nextData` instead of accidentally preserving "old data" that shouldn't be part of "next data". I realize we write to `job.Data` to increment the `file_count`, but maybe we could return that from `saveData`, assign to `nextData`, and simplify this?

File: server/channels/jobs/batch_report_worker.go [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+		return appErr
+	}
+
+	if appErr = worker.app.SendReportToUser(rctx, requestingUserId, job.Id, worker.reportFormat); appErr != nil {
+		return appErr
+	}
+
+	go func() {
+		worker.app.CleanupReportChunks(worker.reportFormat, job.Id, fileCount)
+	}()
```
Comment (lieut-data) [inline server/channels/jobs/batch_report_worker.go:146]: If we fail to send above, will we ever clean up these files?

File: server/channels/api4/report.go [hunk 13]
```diff
@@ -85,3 +88,46 @@ func getUsersForReporting(c *Context, w http.ResponseWriter, r *http.Request) {
...
+
+	if err := c.App.StartUsersBatchExport(c.AppContext, r.URL.Query().Get("date_range")); err != nil {
+		c.Err = err
+		return
+	}
+
+	ReturnStatusOK(w)
+}
+
+func retrieveBatchReportFile(c *Context, w http.ResponseWriter, r *http.Request) {
```
Comment (lieut-data) [inline server/channels/api4/report.go:106]: Why do we have a discrete API to fetch this file and not just include it as an attachment to direct message we send the user? Would that also absolve us of the need to maintain a custom post type?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks, @devinbinnie! A handful of comments inline, with one overarching perspective: there's a lot of "pre-emptive" layers here in anticipation of non-CSV export in the future, but I /suspect/ we'll end up changing much of this to support that other format, making the layers here perhaps unnecessary overhead (especially if we don't revisit this anytime soon). Would it be worth slimming things down now instead?

**Comment by johndavidlugtu** [comment]
> Also, I think we should add a couple of rows at the start to include report metadata like who generated it, what params were used, what time period etc, you know, like a bank statement. Where if I send the report, the receiver doesn't need to ask me all that, its just right in there, in the report.

I think we can explore this as a further iteration. For now let's keep it simple and only serve the data from the table.

For future iteration, I'm +1 on serving more info (e.g. in a PDF, or a zip file)

@harshilsharma63 
cc: @abhijit-singh

File: server/channels/api4/report.go [hunk 14]
```diff
@@ -85,3 +88,46 @@ func getUsersForReporting(c *Context, w http.ResponseWriter, r *http.Request) {
 		c.Logger.Warn("Error writing response", mlog.Err(jsonErr))
 	}
 }
+
+func startUsersBatchExport(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !(c.IsSystemAdmin() && c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionSysconsoleReadUserManagementUsers)) {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/api4/report.go:93]: We already do this with all of the other reporting calls so I wanted to keep consistent. I know there's some way you can give users access to the admin console without being a system admin from what I can tell, so I wanted to cover both for this function and decide later if we want to loosen the restriction. On top of that, adding an extra check for security doesn't hurt too much.

File: server/channels/api4/report.go [hunk 15]
```diff
@@ -85,3 +88,46 @@ func getUsersForReporting(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		c.SetPermissionError(model.PermissionSysconsoleReadUserManagementUsers)
+		return
+	}
+
+	if err := c.App.StartUsersBatchExport(c.AppContext, r.URL.Query().Get("date_range")); err != nil {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/api4/report.go:98]: So we can, but we would then need to convert it back to a string afterwards anyways to append it to the job data, so we would be undoing some work.

We could I suppose parse out the timestamps in the API layer, but I think the location of doing that is kind of arbitrary for this case, since the actual timestamps don't need to be validated in the API and they don't mean anything to any logic in the API at this point either.

We could also validate the string at the API layer, but that would also just error out in the job, and would only really only happen if someone called the API directly, and you would still get an error at that point.

File: server/channels/api4/report.go [hunk 16]
```diff
@@ -85,3 +88,46 @@ func getUsersForReporting(c *Context, w http.ResponseWriter, r *http.Request) {
...
+}
+
+func retrieveBatchReportFile(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !(c.IsSystemAdmin() && c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionSysconsoleReadUserManagementUsers)) {
+		c.SetPermissionError(model.PermissionSysconsoleReadUserManagementUsers)
+		return
+	}
+
+	reportId := c.Params.ReportId
+	if reportId == "" || !model.IsValidId(reportId) {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/api4/report.go:113]: Nope, we can change that.

File: server/channels/api4/report.go [hunk 17]
```diff
@@ -85,3 +88,46 @@ func getUsersForReporting(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	file, name, err := c.App.RetrieveBatchReport(reportId, format)
+	if err != nil {
+		c.Err = err
+		return
+	}
+	defer file.Close()
+
+	w.Header().Set("Content-Type", "text/csv") // TODO: other formats
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/api4/report.go:131]: This was a TODO for me, more to mark that we would be setting something else for different formats.
I'm going to remove the TODO but I'll leave the comment so that we know what to change when implementing other formats.

File: server/channels/app/report.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/report.go:172]: Deduplicating makes sense since a) we wouldn't have to overload the server with a bunch of redundant requests (eg. someone hits the Generate button 50 times) and b) there's no point in generating the same report more than once. I suppose there's a case where a user gets added and we'd want to regenerate, but I think with the right messaging we can let the user know that the system won't do that until the first one is done. At least in this case they would only have to wait for one report.

File: server/channels/app/report.go (hunk 5 above)
Comment (devinbinnie) [reply server/channels/app/report.go:23]: I'll address the overarching theme here then.

I think regardless of what the report format is, as we discussed before we wanted to design a system that's going to allow for a resumable batch job flow. To me that will always, in some capacity follow this workflow:
1) Start a batch job
2) Load some piece of the data for the report
3) Generate a piece of the report, save the file
4) Check if there is more data, if so go to step 2
5) If not, proceed to compile all of the pieces
6) Send the completed piece to the user

Now of course, I can't argue that other formats may have some other small nuances that may not EXACTLY fit this model, but based on what I've understood about reporting in the past, most of them will. PDFs for example, would likely be generated page-by-page, and then sewn together at the end. Same would likely go for DOCX, perhaps LaTeX as well. XLS is similar to CSV, so it would also follow the same flow.

I could continue with other formats, but I think it's important to make an effort to define a "line of best fit" here based on the information we have. I would like further reporting work to follow this flow, until there is a compelling reason to change it. And there is a possibility that will happen, but this to me feels like the best way to start.

As for the specific case above, we will always want to translate the data we have into a human readable form of some kind, which would likely be some arbitrary string, formatted as we see fit. I've added the hook per format as we may want to change that based on how we export. Perhaps the best use case for PDF would be a proper human-readable date string, whereas CSV or XLS would be best stored as something read best by MS Excel or other spreadsheet applications.

tl;dr I'm 4/5 on keeping the current formatting, as I believe this is likely the best way to handle report generation for almost all cases.

File: server/channels/jobs/batch_report_worker.go (hunk 7 above)
Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:25]: I was experimenting with using generics before, must have left this one. Will remove.

Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:25]: We allow jobs that use the `batch_report_worker` to add their own methods to the interface, since it appears we define these interfaces separately.

To be honest, I was always kind of confused why we did that to begin with, and not just use `AppIFace` everywhere.

What is the actual correct way of doing this such that other jobs can still use the passed in `app` parameter here?

File: server/channels/jobs/batch_report_worker.go (hunk 8 above)
Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:48]: The `BatchWorker` module requires a `doBatch` parameter to call since it handles the actual batching calls. This is provided by the `BatchReportWorker` struct, so we have to attach it after.

I can be a bit clearer with that and set it separately instead of the whole thing.

File: server/channels/jobs/batch_report_worker.go (hunk 10 above)
Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:90]: That's a good question. When writing this I realized there could be a case for displaying progress, but with the current design there's nowhere to put it, so I hadn't added it.

I'll remove the TODO but leave a comment explaining why it's 0 and that maybe we can revisit later.

File: server/channels/jobs/batch_report_worker.go (hunk 11 above)
Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:88]: I did it to avoid having to force every job to rewrite its own initial metadata (eg. the date range/other filters of the report, which user requested it) on every call of `getData` because I wanted to simplify for those writing jobs that use this architecture.

I'd rather have the jobs have more control over what goes in and out of their metadata instead of trying to be too prescriptive here.

File: server/channels/jobs/batch_report_worker.go (hunk 12 above)
Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:146]: Good point. This should probably be a deferred function that we add after the compilation.
One thing to maybe consider going forward is a method of cleaning out the file store if there are other failed reports.

File: server/channels/api4/report.go (hunk 13 above)
Comment (lieut-data) [reply server/channels/api4/report.go:93]: Could we use attachment permalinks here to share with other admins instead? 
CleanShot 2024-01-11 at 13 18 48@2x

Comment (devinbinnie) [reply server/channels/api4/report.go:93]: Does that not make it public then? That's something we want to avoid I assume right now.

File: server/channels/api4/report.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply server/channels/api4/report.go:80]: > so I wanted to cover both for this function and decide later if we want to loosen the restriction.

At a minimum, let's write the permission check to reflect the current requirements -- which is being a system admin -- and add code later to check for a specific permission if we want to loosen the restriction. Or better yet, introduce a new permission to cover this case and only grant it to system administrators for now (leaving it up to advanced permissions to customize).

What we have right now really muddies the permission check, making it hard to read and immediately grasp what's going on, since the `model.PermissionSysconsoleReadUserManagementUsers` is illusory.

File: server/channels/api4/report.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply server/channels/api4/report.go:85]: > So we can, but we would then need to convert it back to a string afterwards anyways to append it to the job data, so we would be undoing some work.

I actually think this would be a net win -- we wouldn't ideally be persisting raw data from the enduser as job metadata anyway. Besides any unexpected security concerns, this decouples the "schema" of this parameter between the two use cases: how we accept user input, and how we store offset metadata in the job database. One could imagine a future scenario where we want to programmatically generate reports, and be forced to stitch together a string to communicate this information instead of passing it explicitly.

File: server/channels/api4/report.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply server/channels/api4/report.go]: >  I'm going to remove the TODO but I'll leave the comment so that we know what to change when implementing other formats.

nit: It really feels like it would be obvious what to chagne when implementing other formats, so I'd be inclined to narrow in on CSV vs. future proofing comments.

File: server/channels/app/report.go (hunk 5 above)
Comment (lieut-data) [reply server/channels/app/report.go:23]: Do you have an estimate for how soon we expect to start supporting other formats? CSV being what it is, I'm wondering if shipping CSV now and for the forseeable future would more than cover the required customer scenarios.

I'm only 1/5 on this given the code exists for now, but I'm always hesitant seeing preemptive complexity when N=1.

File: server/channels/jobs/batch_report_worker.go (hunk 7 above)
Comment (lieut-data) [reply server/channels/jobs/batch_report_worker.go:25]: I'd be fine with `AppIFace` over generics in this case, but if we wanted to avoid that, I suppose the more specific worker would need to define its own `app` and interface which is just a superset of the one this worker requires (and for which it maintains its own `app` as well).

File: server/channels/jobs/batch_report_worker.go (hunk 11 above)
Comment (lieut-data) [reply server/channels/jobs/batch_report_worker.go:88]: > I'd rather have the jobs have more control over what goes in and out of their metadata instead of trying to be too prescriptive here.

Agree with this sentiment, but then wouldn't that mean doing `job.Data = nextData`, and letting jobs have more control? As it stands now, we're assuming they want to preserve old job metadata, but this worker can't know that without being too prescriptive.

File: server/channels/api4/report.go (hunk 13 above)
Comment (lieut-data) [reply server/channels/api4/report.go:93]: Yeah, fair enough. I'd be curious if we can reuse the same file info in multiple channels.

File: server/channels/app/report.go (hunk 5 above)
Comment (devinbinnie) [reply server/channels/app/report.go:23]: So phase 1 of this work was only to support CSV, but the original proposal was to include at least PDF reporting. I don't believe following phases have been scheduled yet, but I assume they will be coming down the pipeline, and possibly sooner once the feature goes live, as I can see customers wanting something easier to visualize for non-technical users that comes out "ready-to-go".

Additionally, I'll reiterate that I think it's important to set up a system for how we'd want this to work going forward when we support other formats as I would assume it will be "when", not "if" we do. So I'd like to keep this design as it is to be prescriptive about what it means to add another report format (ie. add the case statements, add the hooks, create a job and go)

File: server/channels/jobs/batch_report_worker.go (hunk 7 above)
Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:25]: Let me try and make that work, as I think that is cleaner. As an aside, I really like generics :P

File: server/channels/jobs/batch_report_worker.go (hunk 11 above)
Comment (devinbinnie) [reply server/channels/jobs/batch_report_worker.go:88]: Hmm I suppose that's true, I didn't consider the case of wanting to have the job completely "unset" a piece of data. I think that's a valid case to change it, so I'll update this.

File: server/channels/api4/report.go (hunk 13 above)
Comment (devinbinnie) [reply server/channels/api4/report.go:93]: I think the easiest way is to just have admins share the link for now, but as suggested above by Harshil we may want to have some system to distribute to multiple users, but that's a future concern for now.

File: server/channels/api4/report.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/api4/report.go]: I'll remove the comment altogether then if you feel it's getting in the way.

File: server/channels/api4/report.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/api4/report.go:85]: Okay, I would say it still seems a bit redundant, but from an architecture perspective it does make sense to always have some kind of "translation" happening between the API and the app logic, so the redundancy will apply in some cases.

I'll change it to compute the date range here.

**Comment by devinbinnie** [comment]
I'm also working on adding unit tests for this change. Will re-request review once those are added as well.

File: server/channels/jobs/batch_migration_worker_test.go [hunk 18]
```diff
@@ -288,6 +288,7 @@ func TestBatchMigrationWorker(t *testing.T) {
 			require.Equal(t, 1, batchNumber, "only batch 1 should have run")
 
 			if batchNumber >= 1 {
+				go worker.Stop()
```
Comment (devinbinnie) [inline server/channels/jobs/batch_migration_worker_test.go:291]: @lieut-data I made some changes to the tests because I noticed that we weren't actually running the correct number of batches in a couple cases, but it also revealed that for some reason the worker won't stop running unless we explicitly run `worker.Stop()`.

In these tests, this shouldn't be required since the job server should stop them once we tell the batch migration worker that it's done, but a later cleanup step fails without these. Is this just a limitation of the tests (ie. the job server is supposed to do this) or do we have an issue with the batch worker?

Comment (lieut-data) [reply server/channels/jobs/batch_migration_worker_test.go:291]: > it also revealed that for some reason the worker won't stop running unless we explicitly run worker.Stop().

I think this is expected, given the worker's Run loop either waiting for new jobs or the stop signal:
https://github.com/mattermost/mattermost/blob/f72aed2263a986becf69fe9a98cccc1e40c7d0cd/server/channels/jobs/batch_migration_worker.go#L78-L87

> In these tests, this shouldn't be required since the job server should stop them once we tell the batch migration worker that it's done

I'm not sure this is true: isn't a worker supposed to remain alive indefinitely to process a new job, only shutting down when the server stops?

> but a later cleanup step fails without these

Can you elaborate on what exactly is failing here?

Comment (devinbinnie) [reply server/channels/jobs/batch_migration_worker_test.go:291]: Okay so I am misunderstanding the concept of the "worker" here, I think I'm misconstruing it with an actual "job". So what should be happening then is that the worker should only be shutting down when sent the `stop` signal, meaning that I think we need to remove the `waitDone` check for every single test here, since not all of them should be shutting down the worker.

**Comment by devinbinnie** [comment]
@lieut-data Ready for re-review. Heads up, I've reworked a bunch of the tests for the batch workers to better suit testing each modules features. Let me know if anything seems particularly off.

File: server/channels/jobs/batch_migration_worker_test.go (hunk 18 above)
Comment (lieut-data) [reply server/channels/jobs/batch_migration_worker_test.go]: > So what should be happening then is that the worker should only be shutting down when sent the stop signal, meaning that I think we need to remove the `waitDone` check for every single test here, since not all of them should be shutting down the worker.

I might be misremembering, but I believe `waitDone` is being used to assert the total behaviour of the worker by asserting that it stops vs. accidentally passing a test while the worker actually plods on ahead and tries to do more on the job than would be expected. Unless there's something demonstrably wrong, I think `waitDone` is working as designed.

Comment (devinbinnie) [reply server/channels/jobs/batch_migration_worker_test.go]: Yes I ended up keeping it as is so that should clear it all up.

File: server/channels/app/report.go (hunk 5 above)
Comment (lieut-data) [reply server/channels/app/report.go:24]: > when we support other formats as I would assume it will be "when", not "if" we do

No objection to keeping thing as is, but my assumption would be "if" not "when". Plus, if we could export the raw data, we could easily build additional tools to turn that into different kinds of reports vs. having to extend the core product all the time.

**Review by lieut-data — APPROVED** [review]
Thanks, @devinbinnie! Just a few additional comments, but nothing blocking another review IMO.

File: server/channels/jobs/export_users_to_csv/export_users_to_csv.go [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+const (
+	timeBetweenBatches = 1 * time.Second
+)
+
+type ExportUsersToCSVAppIFace interface {
+	jobs.BatchReportWorkerAppIFace
+	GetUsersForReporting(filter *model.UserReportOptions) ([]*model.UserReport, *model.AppError)
+}
+
+// MakeWorker creates a batch migration worker to delete empty drafts.
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/jobs/export_users_to_csv/export_users_to_csv.go:25]: This is what happens when I start with such a good template ;)

File: server/channels/app/report.go (hunk 5 above)
Comment (devinbinnie) [reply server/channels/app/report.go:24]: I think that might end up being a valid case, but I know the core reason we started this project was to provide a lot of those tools as a part of the core product since customers were already doing a lot of the reporting themselves, which is why I made my assumption. Direction may change, but I guess we'll see what happens.

File: server/channels/jobs/batch_migration_worker_test.go [hunk 20]
```diff
@@ -45,51 +45,18 @@ func (ma *MockApp) SetOutOfSync() {
...
-				return false
-			}
-		}, 5*time.Second, 100*time.Millisecond, msg)
-	}
-
 	setupBatchWorker := func(t *testing.T, th *TestHelper, mockApp *MockApp, doMigrationBatch func(model.StringMap, store.Store) (model.StringMap, bool, error)) (model.Worker, *model.Job) {
 		t.Helper()
 
-		migrationKey := model.NewId()
-		timeBetweenBatches := 1 * time.Second
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/jobs/batch_migration_worker_test.go:65]: Yeah I'll leave them as-is for clarity for now. I think it's fine to have them as such.
